### PR TITLE
DEV-3009 | missing values for recurring contributions in csv export

### PR DESCRIPTION
**Pull request by benjaminEwhite**
Please fill out each section even if it's just with "N/A"

#### What's this PR do?

- Updates approach to testing CSV export data
  - Provides several new [Factory Boy traits](https://factoryboy.readthedocs.io/en/stable/reference.html#traits) that represent typical types of contributions, inclusive of their metadata and Stripe data values.  
  - This and additional changes to `ContributionFactory` required small updates in a relatively large number of places.
  - Tests now do round trip contributions -> csv -> import CSV -> compare with contributions, and ensure we have each of main relevant contribution states (by using new traits).
  - Identified some tech debt in `Contribution.save` method where sometimes Stripe data is fetched (if `contribution.provider_payment_method_id` is not null plus other conditions). This required "papering over" with mocks and monkeypatching in a decent number of tests.
  - Adds several new properties to `Contribution` that get used in CSV export. 

#### Why are we doing this? How does it help us?

Makes contribution CSV export feature usable

#### How should this be manually tested? Please include detailed step-by-step instructions.

In the review app, do an export. There should be full data for both one-time and recurring contributions.

#### Did this PR make changes to the user interface that may require changes to the user-facing docs?

No

#### Have automated unit tests been added? If not, why?

Yes

#### How should this change be communicated to end users?

N/A

#### Are there any smells or added technical debt to note?

Brought some tech debt into relief which was captured in [DEV-3026](https://news-revenue-hub.atlassian.net/browse/DEV-3026)

#### Has this been documented? If so, where?

No

#### What are the relevant tickets? Add a link to any relevant ones.

[DEV-3009](https://news-revenue-hub.atlassian.net/browse/DEV-3009)

#### Do any changes need to be made before deployment to production (adding environment variables, for example)?

No

#### Are there next steps? If so, what? Have tickets been opened for them? List next-step tickets here:

[DEV-3026](https://news-revenue-hub.atlassian.net/browse/DEV-3026)

### CODE REVIEW

File: apps/contributions/tests/factories.py [hunk 1]
```diff
@@ -59,14 +78,64 @@ class Meta:
...
+            "agreed_to_pay_fees": True,
+            "donor_selected_amount": obj.amount / 100,
+            "revenue_program_id": rp.id or "",
+            "revenue_program_slug": rp.slug or "",
+            "referer": settings.SITE_URL,
+        } | kwargs
+        obj.save()
+
+    class Params:
+        # Add explanatory note
```
Comment (njharman) [inline apps/contributions/tests/factories.py:114]: Doing that which this comment requests, would be ideal. Ditto for others.

Comment (njharman) [reply apps/contributions/tests/factories.py:114]: Trait's seem useful, maybe we should do more.  Maybe you could discuss briefly during sprint retro, how/when to use them.

**Review by njharman — CHANGES_REQUESTED** [review]
Looks good, little hard to follow, but that is nature of large touch lots of things.

request to expound on the factory Traits.

File: apps/contributions/tests/factories.py (hunk 1 above)
Comment (benjaminEwhite) [reply apps/contributions/tests/factories.py:114]: yes! missed that in cleanup.

Comment (benjaminEwhite) [reply apps/contributions/tests/factories.py]: 9eecf17

**Review by njharman — APPROVED** [review]

**Review by x110dc — APPROVED** [review]
